### PR TITLE
[Graphbolt] Improve Labor performance

### PR DESCRIPTION
**Pull request by mfbalin**
#6152 showed less than ideal speedups for labor for the G1 test case. Speedup of Neighbor was up to 36x but Labor was limited at 9x. I suspect this is because of the calls to torch tensor allocation calls. So this is my attempt at reducing those calls by utilizing the stack for small memory allocations.

@RamonZhou how does the labor performance look after this change?

### CODE REVIEW

**Comment by mfbalin** [comment]
@Rhett-Ying The MSVC version in CI doesn't seem to play well with modern C++ features. I recommend updating the MSVC version in CI.

File: graphbolt/src/csc_sampling_graph.cc [hunk 1]
```diff
@@ -781,10 +782,16 @@ inline int64_t LaborPick(
     std::iota(picked_data_ptr, picked_data_ptr + num_neighbors, offset);
     return num_neighbors;
   }
-  torch::Tensor heap_tensor = torch::empty({fanout * 2}, torch::kInt32);
+#define STACK_SIZE_LABOR 1024
```
Comment (frozenbugs) [inline graphbolt/src/csc_sampling_graph.cc:785]:
```suggestion
  static const int kLaborStackSize = 1024;
```

File: graphbolt/src/csc_sampling_graph.cc [hunk 2]
```diff
@@ -781,10 +782,16 @@ inline int64_t LaborPick(
...
   // Assuming max_degree of a vertex is <= 4 billion.
-  auto heap_data = reinterpret_cast<std::pair<float, uint32_t>*>(
-      heap_tensor.data_ptr<int32_t>());
+  std::array<std::pair<float, uint32_t>, STACK_SIZE_LABOR> heap_stack;
```
Comment (frozenbugs) [inline graphbolt/src/csc_sampling_graph.cc:787]: just stack? heap? it is a little bit confusing. Maybe a more descriptive name?

File: graphbolt/src/csc_sampling_graph.cc [hunk 3]
```diff
@@ -814,9 +821,14 @@ inline int64_t LaborPick(
           // is O((fanout + num_neighbors) log(fanout)). It is possible to
           // decrease the logarithmic factor down to
           // O(log(min(fanout, num_neighbors))).
-          torch::Tensor remaining =
-              torch::ones({num_neighbors}, torch::kFloat32);
-          float* rem_data = remaining.data_ptr<float>();
+          std::array<float, STACK_SIZE_LABOR> remaining_stack;
+          auto rem_data = remaining_stack.data();
```
Comment (frozenbugs) [inline graphbolt/src/csc_sampling_graph.cc:825]:
```suggestion
          auto remaining_data = remaining_stack.data();
```

unnecessary abbr.

**Review by frozenbugs — APPROVED** [review]

**Comment by RamonZhou** [comment]
I've tested the performance.

- `v0`: The original one.
- `v1`: After #6152 
- `v2`: After this PR.

L-UF
L-NF
L-UT
L-NT

Data:
data

There is a big boost in G1.
The small drops in other graphs may be fluctuations.

File: graphbolt/src/csc_sampling_graph.cc (hunk 1 above)
Comment (mfbalin) [reply graphbolt/src/csc_sampling_graph.cc:785]: I originally defined it as constexpr int, but msvc in CI doesn't support it properly. It is used as a template parameter for std::array so it needs to be defined at compile time.

File: graphbolt/src/csc_sampling_graph.cc (hunk 2 above)
Comment (mfbalin) [reply graphbolt/src/csc_sampling_graph.cc:787]: Storage is from stack but the datastructure is heap. Maybe heap_from_stack?

Comment (frozenbugs) [reply graphbolt/src/csc_sampling_graph.cc:787]: Just heap should be enough, and L788 makes more sense `heap_data = heap.data();`

File: graphbolt/src/csc_sampling_graph.cc (hunk 1 above)
Comment (frozenbugs) [reply graphbolt/src/csc_sampling_graph.cc:785]: Weird, if that's the case, move the `#define` to the top, after L23. 

`#define kLaborStackSize  1024`

**Comment by mfbalin** [comment]
@frozenbugs My goal is to keep Labor performance at parity with Neighbor as much as possible. Labor actually does a bit more work, always O(num_neighbors) compared to Neighbor possibly doing O(fanout), so it may not be possible for all use cases (high average degree graphs). Since I am proposing Labor to replace Neighbor, I don't want potential future users to say this is a complicated algorithm so it is bound to be slow. They keep saying this in my paper reviews all the time.

What would be nice is to give this algorithm a bit more visibility in examples etc. As it is, the API and behavior fully match the Neighbor API so one can go ahead and use this instead of Neighbor in some examples. I would love the feedback if somehow the end accuracy or result doesn't match the Neighbor baseline, which I could then investigate, but I don't expect such a thing to happen. When sampling multiple layers, Labor will sample significantly fewer vertices and edges which will speed up convergence.

In the future, Labor will support more advanced features which are already supported by the `dgl.sampling.sample_labors`. I am sure it will all look very fancy when we announce all these supported features to the wider GNN community.

File: graphbolt/include/graphbolt/csc_sampling_graph.h [hunk 4]
```diff
@@ -406,8 +406,8 @@ int64_t PickByEtype(
     PickedType* picked_data_ptr);
 
 template <
-    bool NonUniform, bool Replace, typename ProbsType = float,
-    typename PickedType>
+    bool NonUniform, bool Replace, typename ProbsType, typename PickedType,
```
Comment (mfbalin) [inline graphbolt/include/graphbolt/csc_sampling_graph.h:409]: C++ requires template parameters with default values to come after the other template parameters. So, it is necessary to delete the default from ProbsType. I don't know how it was compiling before.

File: graphbolt/src/csc_sampling_graph.cc [hunk 5]
```diff
@@ -781,10 +783,15 @@ inline int64_t LaborPick(
     std::iota(picked_data_ptr, picked_data_ptr + num_neighbors, offset);
     return num_neighbors;
   }
-  torch::Tensor heap_tensor = torch::empty({fanout * 2}, torch::kInt32);
   // Assuming max_degree of a vertex is <= 4 billion.
-  auto heap_data = reinterpret_cast<std::pair<float, uint32_t>*>(
-      heap_tensor.data_ptr<int32_t>());
+  std::array<std::pair<float, uint32_t>, StackSize> heap;
```
Comment (Rhett-Ying) [inline graphbolt/src/csc_sampling_graph.cc:787]: so the main change here is use stack memory instead of heap memory for small size to achieve the improvement?

File: graphbolt/src/csc_sampling_graph.cc [hunk 6]
```diff
@@ -781,10 +783,15 @@ inline int64_t LaborPick(
...
+  auto heap_data = heap.data();
+  torch::Tensor heap_tensor;
+  if (fanout > StackSize) {
+    heap_tensor = torch::empty({fanout * 2}, torch::kInt32);
```
Comment (Rhett-Ying) [inline graphbolt/src/csc_sampling_graph.cc:791]: Is this memory safe? `heap_tensor` holds the memory and it ends it's life outside of the `if` statement. Such memory will be deleted automatically even the data pointer is saved to `heap_data`?

File: graphbolt/src/csc_sampling_graph.cc [hunk 7]
```diff
@@ -781,10 +783,15 @@ inline int64_t LaborPick(
...
+    heap_data = reinterpret_cast<std::pair<float, uint32_t>*>(
```
Comment (Rhett-Ying) [inline graphbolt/src/csc_sampling_graph.cc:792]: Are we assuming the memory size allocated by `torch::empty` equals `fanout * 2 * sizeof(float) `?

File: graphbolt/src/csc_sampling_graph.cc [hunk 8]
```diff
@@ -781,10 +783,15 @@ inline int64_t LaborPick(
...
+        heap_tensor.data_ptr<int32_t>());
```
Comment (Rhett-Ying) [inline graphbolt/src/csc_sampling_graph.cc:793]: `float`, `uint32_t`, `int32_t` are used in different places while we always assume all of them are in same size, namely `4` bytes. We may need a more secure way to violate it if someone else change one of them to different size such as `int64_t` some day.

File: graphbolt/src/csc_sampling_graph.cc (hunk 5 above)
Comment (mfbalin) [reply graphbolt/src/csc_sampling_graph.cc:787]: Yes

File: graphbolt/src/csc_sampling_graph.cc (hunk 6 above)
Comment (mfbalin) [reply graphbolt/src/csc_sampling_graph.cc:791]: heap_tensor is defined outside of if.

File: graphbolt/src/csc_sampling_graph.cc (hunk 7 above)
Comment (mfbalin) [reply graphbolt/src/csc_sampling_graph.cc:792]: we are assuming it equals fanout int32_t and fanout floats.

File: graphbolt/src/csc_sampling_graph.cc (hunk 8 above)
Comment (mfbalin) [reply graphbolt/src/csc_sampling_graph.cc:793]: This is already secure. My only assumption is that a vertex doesn't have more than 4 billion neighbors. There is no dataset that I know where a vertex has more than 4 billion neighbors.

Comment (mfbalin) [reply graphbolt/src/csc_sampling_graph.cc:793]: I can change the code so that it allocates a tensor with enough bytes in guaranteed manner even if the types change. Let me do that.

File: graphbolt/src/csc_sampling_graph.cc (hunk 6 above)
Comment (Rhett-Ying) [reply graphbolt/src/csc_sampling_graph.cc:791]: utilizing `torch:empty` instead of `new` or `std::vector` is to ask torch to help us manage memory better?

Comment (mfbalin) [reply graphbolt/src/csc_sampling_graph.cc]: Yeah, I thought that since we are relying on torch for everything, we can do the same here. Not sure if torch allocation is faster or new is faster.

File: graphbolt/src/csc_sampling_graph.cc (hunk 8 above)
Comment (mfbalin) [reply graphbolt/src/csc_sampling_graph.cc:794]: Just made a change.

**Review by Rhett-Ying — APPROVED** [review]

**Comment by Rhett-Ying** [comment]
> @Rhett-Ying The MSVC version in CI doesn't seem to play well with modern C++ features. I recommend updating the MSVC version in CI.

it's VS2019 and may not play well with C++17 features.

File: graphbolt/src/csc_sampling_graph.cc [hunk 9]
```diff
@@ -781,10 +783,16 @@ inline int64_t LaborPick(
...
-  torch::Tensor heap_tensor = torch::empty({fanout * 2}, torch::kInt32);
   // Assuming max_degree of a vertex is <= 4 billion.
-  auto heap_data = reinterpret_cast<std::pair<float, uint32_t>*>(
-      heap_tensor.data_ptr<int32_t>());
+  std::array<std::pair<float, uint32_t>, StackSize> heap;
+  auto heap_data = heap.data();
+  torch::Tensor heap_tensor;
+  if (fanout > StackSize) {
+    heap_tensor = torch::empty(
+        {fanout * (sizeof(heap_data[0]) / sizeof(int32_t))}, torch::kInt32);
```
Comment (Rhett-Ying) [inline graphbolt/src/csc_sampling_graph.cc]: seems cause error when building on windows, `2` is ok for me as we explicitly declare `std::pair` above. Or try with `std::tuple_size<decltype(heap)::value_type>::value`?

Comment (mfbalin) [reply graphbolt/src/csc_sampling_graph.cc]: I just fixed it. It was related to casting of unsigned int to int.

**Comment by mfbalin** [comment]
We also could allocate the required memory inside the parallel for section but that seemed like a lot of code change. I didn't want to make the code more complicated so I resorted to using the stack as a quick hack.

**Comment by mfbalin** [comment]
I am pretty happy with the current performance of this code. Made a last-minute minor optimization. Feel free to merge anytime.
ogbn-products dataset 20 fanout, 200k nodes:
```
{('NS', False, False): 13.961718496284448,
 ('LABOR', False, False): 22.257723700022325,
 ('NS', False, True): 264.3265006976435,
 ('LABOR', False, True): 27.74477390339598,
 ('NS', True, False): 198.85092989716213,
 ('LABOR', True, False): 47.24292550235987,
 ('NS', True, True): 1236.366263902164,
 ('LABOR', True, True): 54.78198000055272}
```
Benchmarked with following code:
```
import torch
import dgl.graphbolt as gb
from ogb.nodeproppred import DglNodePropPredDataset
from dgl.data import RedditDataset
# g = gb.from_dglgraph(RedditDataset()[0])
g = gb.from_dglgraph(DglNodePropPredDataset('ogbn-products')[0][0])
g = gb.from_csc(g.csc_indptr, g.indices, None, None, {'probs': torch.rand([g.num_edges])}, g.metadata)
from timeit import default_timer as timer
from collections import defaultdict
times = defaultdict(float)
trials = 10
for z in range(trials):
    nodes = torch.randperm(g.num_nodes)[:200000]
    # nodes = torch.arange(g.num_nodes)
    for replace in [False, True]:
        for w in [None, 'probs']:
            for f, name in zip([g.sample_neighbors, g.sample_layer_neighbors], ["NS", "LABOR"]):
                st = timer()
                out = f(nodes, torch.tensor([20]), replace, probs_name=w)
                t = timer() - st
                times[name, replace, w is not None] += t
print({k: v / trials * 1000 for k, v in times.items()})
```

File: graphbolt/src/csc_sampling_graph.cc [hunk 10]
```diff
@@ -781,10 +783,16 @@ inline int64_t LaborPick(
...
     return num_neighbors;
   }
-  torch::Tensor heap_tensor = torch::empty({fanout * 2}, torch::kInt32);
   // Assuming max_degree of a vertex is <= 4 billion.
-  auto heap_data = reinterpret_cast<std::pair<float, uint32_t>*>(
-      heap_tensor.data_ptr<int32_t>());
+  std::array<std::pair<float, uint32_t>, StackSize> heap;
+  auto heap_data = heap.data();
+  torch::Tensor heap_tensor;
+  if (fanout > StackSize) {
```
Comment (peizhou001) [inline graphbolt/src/csc_sampling_graph.cc:790]: Is that means `fanout > StackSize`, memory allocated in Line 787 is wasted?

File: graphbolt/src/csc_sampling_graph.cc [hunk 11]
```diff
@@ -814,41 +822,62 @@ inline int64_t LaborPick(
...
-              torch::ones({num_neighbors}, torch::kFloat32);
-          float* rem_data = remaining.data_ptr<float>();
+          std::array<float, StackSize> remaining;
+          auto remaining_data = remaining.data();
+          torch::Tensor remaining_tensor;
+          if (num_neighbors > StackSize) {
+            remaining_tensor = torch::empty({num_neighbors}, torch::kFloat32);
+            remaining_data = remaining_tensor.data_ptr<float>();
+          }
+          std::fill(remaining_data, remaining_data + num_neighbors, 1.f);
```
Comment (peizhou001) [inline graphbolt/src/csc_sampling_graph.cc:832]: Will there have some gains to use memset?

File: graphbolt/src/csc_sampling_graph.cc [hunk 12]
```diff
@@ -814,41 +822,62 @@ inline int64_t LaborPick(
...
           // decrease the logarithmic factor down to
           // O(log(min(fanout, num_neighbors))).
-          torch::Tensor remaining =
-              torch::ones({num_neighbors}, torch::kFloat32);
-          float* rem_data = remaining.data_ptr<float>();
+          std::array<float, StackSize> remaining;
+          auto remaining_data = remaining.data();
+          torch::Tensor remaining_tensor;
+          if (num_neighbors > StackSize) {
+            remaining_tensor = torch::empty({num_neighbors}, torch::kFloat32);
```
Comment (peizhou001) [inline graphbolt/src/csc_sampling_graph.cc:829]: Ditto.

File: graphbolt/src/csc_sampling_graph.cc (hunk 11 above)
Comment (RamonZhou) [reply graphbolt/src/csc_sampling_graph.cc:832]: `memset` seems not able to do the same thing here. It fills the memory by bytes.

File: graphbolt/src/csc_sampling_graph.cc (hunk 12 above)
Comment (mfbalin) [reply graphbolt/src/csc_sampling_graph.cc:829]: What do you mean here?

**Comment by mfbalin** [comment]
Can a significant portion of edge weights be zero? Should I optimize considering this case as well in the future? Will users use edge weights to separate train, validation or test edges etc.?

**Comment by frozenbugs** [comment]
> @frozenbugs @Rhett-Ying @peizhou001 Can a significant portion of edge weights be zero? Should I optimize considering this case as well in the future? Will users use edge weights to separate train, validation or test edges etc. by setting a lot of edge weights to 0?

It is hard to predict how users will use the edge weight in a very specific way, but we usually only optimize for common usage, if a user pattern become a common pattern, we can always optimize it later.

**Review by peizhou001 — APPROVED** [review]
